draft-wang-ppm-dap-taskprov.md

```diff
@@ -56,69 +60,304 @@ collector). Thus it is up to each deployment of DAP to securely implement its
+    task-prov(0xff00),
+    (65535)
 } ExtensionType;
 ~~~
+> NOTE(shan) TLS uses underscore `_` to define codepoints, which is different
+> from DAP's convension of using hyphen `-`, which one to follow?
+> See [#314](ietf-wg-ppm/draft-ietf-ppm-dap#314) for discussion.
 
 When the Client includes this extension with its report, the body of the
 extension is structured as follows:
 
 ~~~
-TODO(wangshan) Specify the structure of the extension.
+struct {
+    /* Info specific for a task. */
+    opaque task_info<1..2^8-1>;
+
+    /* A list of URLs relative to which an aggregator's API endpoints */
+    /* can be found. */
+    Url aggregator_endpoints<1..2^16-1>; // Defined in I-D.draft-ietf-ppm-dap-02
+
+    /* This determines the query type for batch selection and the */
+    /* properties that all batches for this task must have. */
+    QueryConfig query_config; // Defined in I-D.draft-ietf-ppm-dap-02
+
+    /* The maximum number of times a batch of reports may be queried */
+    /* by the Collector. */
+    uint16 max_batch_lifetime;
+
+    /* Time up to which clients are allowed to upload to this task. */
+    /* See https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/304 */
+    Time task_expiration; // Defined in I-D.draft-ietf-ppm-dap-02
+
+    /* A unique identifier for the VDAF instance used for the task, */
+    /* including the type of measurement associated with the task. */
+    VdafType vdaf_type;
+
+    /* Additional parameters relevant for the vdaf_type */
+    opaque vdaf_data<1..2^16-1>;
+} TaskConfig;
+
+struct {
+    uint32 max_batch_size;
+} FixedSizeQueryConfig;
+
+struct {
+    QueryType query_type; // Defined in I-D.draft-ietf-ppm-dap-02
+    Duration time_precision; // Defined in I-D.draft-ietf-ppm-dap-01
+    uint32 min_batch_size;
+    select (query_type) {
+        case time-interval: Empty;
+        case fixed-size: FixedSizeQueryConfig fixed_size_query_config;
+    }
+} QueryConfig;
 ~~~
 
+The purpose of `TaskConfig` is to include all parameters that are necessary
+for creating a new task in aggregators. It includes all the fields to be
+associated with a task (see task configuration in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.). Besides, `TaskConfig` also includes fields
+that useful for configuring a task in-band:
+
+* An opaque `task_info` that is specific to a task. For e.g. this can be a
+string describing the purpose of this task. It can also be the DAP task ID
+chosen by task author.
+
+* An opaque `vdaf_data` that contains any VDAF specific parameters for the
+chosen `vdaf_type`. The aggregators MUST pass `vdaf_data` to VDAF initialiser,
+based on the chosen `vdaf_type`.
+
+The definition of Time, Duration, Url, QueryType follow those in
+{{?DAP=I-D.draft-ietf-ppm-dap-01}}.
+
+Note that the parameters that are not necessarily tied to a task (for e.g.
+collector hpke config), and secrets that should not be known by clients, like
+`vdaf_verify_key`, MUST still be exchanged out-of-band among aggregators.
```

Comment:
I don't think this paragraph adds anything. It can be removed.

Reply:
@cjpatton The core protocol task configuration includes secrets like `vdaf_verify_key`, I think this should be mentioned explicitly?  @junyechen1996 your comment https://github.com/wangshan/draft-wang-ppm-dap-taskprov/pull/2/files#r964267738 is basically addressed here

Reply:
Hmm, I don't think my comment was very helpful. Please allow me to try again :)

It's useful to note that TaskConfig does not encode all parameters required to run a task. (Listing them is also helpful: I believe this just includes the VDAF verification key and the collector HPKE config.)

What's less useful here is saying that these MUST be exchanged out-of-band. It's best to reserve the keyword "MUST" for prescriptive things, e.g., "If the client does X, then the server MUST do Y". Here we're saying that there are fields that MUS be exchanged out-of-band, but we don't say how.

I would be happy with something like this: 

> Note that `TaskConfig` does not encode all of the parameters required for the aggregator to run a task. In particular, the VDAF verification key, the collector HPKE configuration, and whatever assets are required for HTTP request authentication are still established out-of-band.

Side note: Is the plan to to derive the VDAF verification key from the a shared secret on-the-fly?

Reply:
I see, let me update it.

Re: vdaf_verify_key, I think our last conclusion was that if the leader chooses vdaf_verify_key, then the privacy leak threat already exists, generate it using a shared secret doesn't help, do we have consensus on this?